driver-core/src/main/com/mongodb/KerberosSubjectProvider.java

```diff
@@ -80,16 +89,32 @@ public KerberosSubjectProvider(@NonNull final String loginContextName) {
     @NonNull
     public synchronized Subject getSubject() throws LoginException {
         if (subject == null || needNewSubject(subject)) {
-            LOGGER.info("Creating new LoginContext and logging in the principal");
-            LoginContext loginContext = new LoginContext(loginContextName);
-            loginContext.login();
-            subject = loginContext.getSubject();
-            LOGGER.info("Login successful");
+            subject = createNewSubject();
         }
         return subject;
     }
 
-    private static boolean needNewSubject(@NonNull final Subject subject) {
+    private Subject createNewSubject() throws LoginException {
+        LoginContext loginContext;
+        try {
+            LOGGER.debug(format("Creating LoginContext with name '%s'", loginContextName));
+            loginContext = new LoginContext(loginContextName);
+        } catch (LoginException e) {
+            if (fallbackLoginContextName == null) {
+                throw e;
+            }
+            LOGGER.debug(format("Creating LoginContext with fallback name '%s'", fallbackLoginContextName));
+            loginContext = new LoginContext(fallbackLoginContextName);
+            loginContextName = fallbackLoginContextName;
```

Comment:
Once we fall back once, always fall back

Reply:
If we want this behavior, then we should execute
```java
loginContextName = fallbackLoginContextName;
fallbackLoginContextName = null;
```
before attempting to construct a `LoginContext` within the `catch` block: if this attempt also fails, then the next execution of the `getSubject`method (they are totally ordered as a result of `synchronized`) uses the initial `loginContextName` instead of using the fallback name.

Reply:
Sorry, I was imprecise.  I meant to say that once we fallback to calling the `_login_` method, then we always fall back.  I don't actually think it matters much what we do if the `LoginContext` constructor throws for both names.